clipper_admin/docs/model_deployers.rst

```diff
@@ -0,0 +1,163 @@
+* ``sc``: The current SparkContext. The PySpark model serialization API requires the SparkContext as an argument
+
+The effect of these two changes is to allow the deployed prediction function to capture all pure Python
+state through closure capture but explicitly declare the additional PySpark state which must be saved and
+loaded through a separate process.
+
+All caveats about dependency-capture still hold.
+
+.. autofunction:: clipper_admin.deployers.pyspark.deploy_pyspark_model
+.. autofunction:: clipper_admin.deployers.pyspark.create_endpoint
+
+
+PyTorch Models
+--------------
+For PyTorch, Clipper will serialize the model using ``torch.save`` and it will be loaded using ``torch.load``. It is expected the model has a forward method and can be called using ``model(input)`` to predict output. 
+
+.. autofunction:: clipper_admin.deployers.pytorch.deploy_pytorch_model
+.. autofunction:: clipper_admin.deployers.pytorch.create_endpoint
+
+
+Tensorflow Models
+-----------------
+For Tensorflow, Clipper will serialize the Tensorflow Session.
+
+.. autofunction:: clipper_admin.deployers.tensorflow.deploy_tensorflow_model
+.. autofunction:: clipper_admin.deployers.tensorflow.create_endpoint
+
+
+MXNet Models
+------------
+For MXNet, Clipper will serialize the model using ``mxnet_model.save_checkpoint(..., epoch=0)``.
+
+.. autofunction:: clipper_admin.deployers.mxnet.deploy_mxnet_model
+.. autofunction:: clipper_admin.deployers.mxnet.create_endpoint
+
+
+PyTorch->ONNX->Caffe2 Models
+----------------------------
+.. warning::
+    ONNX support is experimental. The model container might not be stable. As of 0.3 release, we expect the Caffe2 backend will function with some limitation. 
+
+This module currently takes a PyTorch model, save it to ONNX files, and serve it using Caffe2. 
+
+.. autofunction:: clipper_admin.deployers.onnx.deploy_pytorch_model
+.. autofunction:: clipper_admin.deployers.onnx.create_pytorch_endpoint
+
+
+Create Your Own Container
+-------------------------
+.. note::
+    This method should be considered as the "last resort" when both ``deploy_python_closure`` won't work 
+    due to dependency issue and/or Clipper does not provide pre-built containers. 
+
+You can create a new Docker image based on the default
+image ``clipper/python-closure-container:develop`` that installs these additional dependencies.
+
+For example, if you knew your function needed the Python package ``networkx``, you could create
+the following Dockerfile::
+
+    FROM clipper/python-closure-container:develop
+    RUN pip install networkx
+
+You would then use this Dockerfile to build a new image::
+
+    docker build -t python-closure-with-networkx .
+
+Then, when you call ``deploy_python_closure`` or ``create_endpoint``, instead of
+leaving the ``base_image`` argument to be the default, you would set it to
+``base_image="python-closure-with-networkx"`` and now the container won't need to install
+networkx each time th container is initialized. 
```

Comment:
This example isn't relevant anymore, because you can specify additional pip packages to install.